core-js/src/definitions.ts

```diff
@@ -0,0 +1,1421 @@
+
+export type QueryRoot = 'query';
+export type MutationRoot = 'mutation';
+export type SubscriptionRoot = 'subscription';
+export type SchemaRoot = QueryRoot | MutationRoot | SubscriptionRoot;
+
+export type Type = InputType | OutputType;
+export type NamedType = ScalarType | ObjectType | InterfaceType | UnionType | EnumType | InputObjectType;
+export type OutputType = ScalarType | ObjectType | InterfaceType | UnionType | EnumType | ListType<any> | NonNullType<any>;
+export type InputType = ScalarType | EnumType | InputObjectType | ListType<any> | NonNullType<any>;
+export type WrapperType = ListType<any> | NonNullType<any>;
+
+export type OutputTypeReferencer = FieldDefinition<any>;
+export type InputTypeReferencer = InputFieldDefinition | ArgumentDefinition<any>;
+export type ObjectTypeReferencer = OutputTypeReferencer | UnionType | SchemaDefinition;
+export type InterfaceTypeReferencer = OutputTypeReferencer | ObjectType | InterfaceType;
+
+export type NullableType = NamedType | ListType<any>;
+
+export type NamedTypeKind = NamedType['kind'];
+
+export function defaultRootTypeName(root: SchemaRoot) {
+  return root.charAt(0).toUpperCase() + root.slice(1);
+}
+
+export function isNamedType(type: Type): type is NamedType {
+  return type instanceof BaseNamedType;
+}
+
+export function isWrapperType(type: Type): type is WrapperType {
+  return type.kind == 'ListType' || type.kind == 'NonNullType';
+}
+
+export function isOutputType(type: Type): type is OutputType {
+  if (isWrapperType(type)) {
+    return isOutputType(type.baseType());
+  }
+  switch (type.kind) {
+    case 'ScalarType':
+    case 'ObjectType':
+    case 'UnionType':
+    case 'EnumType':
+    case 'InterfaceType':
+      return true;
+    default:
+      return false;
+  }
+}
+
+export function ensureOutputType(type: Type): OutputType {
+  if (isOutputType(type)) {
+    return type;
+  } else {
+    throw new Error(`Type ${type} (${type.kind}) is not an output type`);
+  }
+}
+
+export function isInputType(type: Type): type is InputType {
+  if (isWrapperType(type)) {
+    return isInputType(type.baseType());
+  }
+  switch (type.kind) {
+    case 'ScalarType':
+    case 'EnumType':
+    case 'InputObjectType':
+      return true;
+    default:
+      return false;
+  }
+}
+
+export function ensureInputType(type: Type): InputType {
+  if (isInputType(type)) {
+    return type;
+  } else {
+    throw new Error(`Type ${type} (${type.kind}) is not an input type`);
+  }
+}
+
+export interface Named {
+  readonly name: string;
+}
+
+export abstract class SchemaElement<Parent extends SchemaElement<any, any> | Schema, Referencer> {
+  protected _parent?: Parent;
+  protected readonly _appliedDirectives: Directive[] = [];
+  description?: string;
+  source?: ASTNode;
+
+  abstract coordinate: string;
+
+  schema(): Schema | undefined {
+    if (!this._parent) {
+      return undefined;
+    } else if ( this._parent instanceof Schema) {
+      return this._parent as any;
```

Comment:
That surprised me as well, but `as Schema` does not work. And what I suspect is that this has to do with some limitation of typescript type system with generics (or I'm being stupid, which is always an option).

Namely, even after the `instanceof`, typescript says that `this._parent` is of type `Parent & Schema`, where `Parent` is the generic argument defined as `Parent extends SchemaElement | Schema`), and complains that this isn't assignable to `Schema | undefined` (and it says it's not because `Type 'Parent & Schema' is missing some properties from type 'Schema'` which does not make sense to me: either the `&` in typescript has a really weird semantic, or it sound like a bug/limitation). But due to that, it also refuse `as Schema` saying that it "may be a mistake" and we should should "convert to 'unknown" first.

Tl;dr, `as any` is the shortest way I know to make it shut-up.

I will definitively add a comment though.